### PR TITLE
line and bar chart combination and enhancements

### PR DESCRIPTION
**Pull request by szaghikian-rasgo**
Initial changes to combine line and bar charts. More experience decisions need to be discussed, including what we want to infer from args passed and what we want to make explicit. This is here as a template/guide.

May include combination of heatmap depending on the direction of the input inference chats.

Looking for specific feedback on:
1. does anyone have a snappy name for what to call a multi-transform combined template?
2. @kimballh usage of the new `get_columns` utility
3. Type comparisons with strings guiding the template flow
4. @phillip-rasgo creating a new chart type (in the YAML file)?

### CODE REVIEW

**Comment by szaghikian-rasgo** [comment]
Testing if you wanna go check it out:
```
#################################################
# RasgoTransforms combine line chart, bar chart, and histogram
#################################################
# ds = rasgo.get.dataset(61)

# Rasgo meta ds
# ds = rasgo.get.dataset(4114)

ds = rasgo.get.dataset(80)

print(ds)

s = """
{%- set axis_type_dict = get_columns(source_table) -%}
{%- set axis_type_response = axis_type_dict[x_axis].upper() -%}
{%- if 'DATE' in axis_type_response or 'TIME' in axis_type_response -%}
  {%- set axis_type = "date" -%}
{%- elif 'NUM' in axis_type_response or 'FLOAT' in axis_type_response or 'INT' in axis_type_response or 'DECIMAL' in axis_type_response or 'DOUBLE' in axis_type_response or 'REAL' in axis_type_response -%}
    {%- set axis_type = "numeric" -%}
{%- elif 'BINARY' in axis_type_response or 'TEXT' in axis_type_response or 'BOOLEAN' in axis_type_response or 'CHAR' in axis_type_response or 'STRING' in axis_type_response or 'VARBINARY' in axis_type_response -%}
    {%- set axis_type = "categorical" -%}
{%- else -%}
    {{ raise_exception('The column selected as an axis is not categorical, numeric, or datetime. Please choose an axis that is any of these data types and recreate the transform.') }}
{%- endif -%}

{%- if num_buckets is not defined -%}
    {%- set bucket_count = 200 -%}
{%- else -%}
    {%- set bucket_count = num_buckets -%}
{%- endif -%}

-- if the axis is continuous or a date, do a line chart
{% if axis_type in ['date', 'numeric'] -%}
    WITH AXIS_RANGE AS (
    -- Use a user-defined axis column to calculate the min & max of the axis (and buckets on the axis)
    SELECT
        {% if axis_type == 'date' -%}
            MIN(DATE_PART(EPOCH_SECOND, {{ x_axis }}))-1 AS MIN_VAL
            ,MAX(DATE_PART(EPOCH_SECOND, {{ x_axis }}))+1 AS MAX_VAL
        {% else %}
            MIN({{ x_axis }})-1 AS MIN_VAL
            ,MAX({{ x_axis }})+1 AS MAX_VAL
        {%- endif %}
    FROM
        {{ source_table }}
    WHERE
        {{ x_axis }} IS NOT NULL
    ), EDGES AS (
    SELECT MIN_VAL, MAX_VAL, (MIN_VAL-MAX_VAL) VAL_RANGE, ((MAX_VAL-MIN_VAL)/{{ bucket_count }}) BUCKET_SIZE FROM AXIS_RANGE
    ),
    BUCKETS AS (
    SELECT
        -- Assigns a bucket to each value of each column in user's column list
        -- Row count of result set should match the row count of the raw table
        MIN_VAL
    ,MAX_VAL
    ,BUCKET_SIZE
    ,{{ "DATE_PART(EPOCH_SECOND, " + x_axis +")" if axis_type == 'date' else x_axis }}::float AS COL_A_VAL
    ,WIDTH_BUCKET(COL_A_VAL, MIN_VAL, MAX_VAL, {{ bucket_count }}) AS COL_A_BUCKET
    {%- for col, aggs in metrics.items() %}
    ,{{ col }}
    {%- endfor %}

    FROM
        {{ source_table }}
        CROSS JOIN EDGES
    )
    -- Run final aggregates on the buckets
    SELECT
    {% if axis_type == 'date' -%}
        (MIN_VAL+((COL_A_BUCKET-1)*BUCKET_SIZE))::DATETIME AS {{ x_axis }}_MIN
        ,(MIN_VAL+(COL_A_BUCKET*BUCKET_SIZE))::DATETIME AS {{ x_axis }}_MAX
    {%- else -%}
        MIN_VAL+((COL_A_BUCKET-1)*BUCKET_SIZE) AS {{ x_axis }}_MIN
        ,MIN_VAL+(COL_A_BUCKET*BUCKET_SIZE) AS {{ x_axis }}_MAX
    {%- endif -%}
    {%- for col, aggs in metrics.items() %}
        {%- for agg in aggs %}
            ,{{ agg }}({{ col }}) AS {{ agg }}_{{ col }}
        {%- endfor -%}
    {%- endfor %}

    FROM BUCKETS
    WHERE {{ x_axis }}_MIN is not NULL
    {%- if filter_statements is iterable -%}
        {%- for filter_statement in filter_statements %}
            AND {{ filter_statement }}
        {%- endfor -%}
    {%- endif %}
    GROUP BY 1, 2
    ORDER BY 1

{%- elif axis_type == 'categorical' -%}
-- if the axis is a categorical dimension, build a bar chart
    SELECT
    {{ x_axis }},

    {%- for col, aggs in metrics.items() %}
        {%- set outer_loop = loop -%}
        {%- for agg in aggs %}
            {{ agg }}({{ col }}) as {{ col + '_' + agg }}{{ '' if loop.last and outer_loop.last else ',' }}
        {%- endfor -%}
    {%- endfor %}
    FROM {{ source_table }}
    {% if filter_statements is iterable -%}
        {%- for filter_statement in filter_statements %}
            {{ 'WHERE' if loop.first else 'AND' }} {{ filter_statement }}
        {%- endfor -%}
    {%- endif %}
    GROUP BY {{ x_axis }}
    {{ "ORDER BY " + x_axis + " " + order_direction if order_direction else '' }}
{%- endif -%}
"""

# Date X-axis with num_buckets set
print(ds.apply(
    sql=s,
    x_axis="DATE",
    metrics={
        "COVID_DEATHS": ["AVG"],
        "COVID_CUMULATIVE_CASES": ["AVG", "SUM"]
    },
    num_buckets=100
).sql)

ds2 = ds.apply(
    sql=s,
    x_axis="DATE",
    metrics={
        "COVID_DEATHS": ["AVG"],
        "COVID_CUMULATIVE_CASES": ["AVG", "SUM"]
    },
    num_buckets=100
)
ds2.preview()

# Date X-axis with num_buckets set and a filter_statement and a order_direction

print(ds.apply(
    sql=s,
    x_axis="DATE",
    metrics={
        "COVID_DEATHS": ["AVG"],
        "COVID_CUMULATIVE_CASES": ["AVG", "SUM"]
    },
    num_buckets=300,
    order_direction="ASC",
    filter_statements=["DATE_MIN BETWEEN '2020-12-01' AND '2020-12-30'"]
).sql)

ds3 = ds.apply(
    sql=s,
    x_axis="DATE",
    metrics={
        "COVID_DEATHS": ["AVG"],
        "COVID_CUMULATIVE_CASES": ["AVG", "SUM"]
    },
    num_buckets=300,
    order_direction="ASC",
    filter_statements=["DATE_MIN BETWEEN '2020-12-01' AND '2020-12-30'"]
)
ds3.preview()

# Numeric X-axis with num buckets and filter statement

print(ds.apply(
    sql=s,
    x_axis="COVID_NEW_CONFIRMED",
    metrics={
        "COVID_DEATHS": ["AVG"],
        "COVID_CUMULATIVE_CASES": ["AVG", "SUM"]
    },
    num_buckets=300,
    filter_statements=["COVID_NEW_CONFIRMED_MIN > 10"]
).sql)

ds4 = ds.apply(
    sql=s,
    x_axis="COVID_NEW_CONFIRMED",
    metrics={
        "COVID_DEATHS": ["AVG"],
        "COVID_CUMULATIVE_CASES": ["AVG", "SUM"]
    },
    num_buckets=300,
    filter_statements=["COVID_NEW_CONFIRMED_MIN > 10"]
)
ds4.preview()

# Numeric X-axis with order statement

print(ds.apply(
    sql=s,
    x_axis="COVID_NEW_CONFIRMED",
    metrics={
        "COVID_DEATHS": ["AVG"],
        "COVID_CUMULATIVE_CASES": ["AVG", "SUM"]
    },
    order_direction="DESC"
).sql)

ds5 = ds.apply(
    sql=s,
    x_axis="COVID_NEW_CONFIRMED",
    metrics={
        "COVID_DEATHS": ["AVG"],
        "COVID_CUMULATIVE_CASES": ["AVG", "SUM"]
    },
    order_direction="DESC"
)
ds5.preview()

# Categorical X-axis, num buckets, filter

ga_ds = rasgo.get.dataset(4566)

print(ga_ds)

print(ga_ds.apply(
    sql=s,
    x_axis="CHANNEL_GROUPING",
    metrics={
        "GOAL_VALUE_ALL": ["SUM"],
        "NEW_USERS": ["AVG", "SUM"]
    },
    num_buckets = 100,
    filter_statements=["CHANNEL_GROUPING IN ('Direct', 'Referral')"]
).sql)

ds6 = ga_ds.apply(
    sql=s,
    x_axis="CHANNEL_GROUPING",
    metrics={
        "GOAL_VALUE_ALL": ["SUM"],
        "NEW_USERS": ["AVG", "SUM"]
    },
    num_buckets = 100,
    filter_statements=["CHANNEL_GROUPING IN ('Direct', 'Referral')"]
)
ds6.preview()

# Categorical X-axis, order

print(ga_ds.apply(
    sql=s,
    x_axis="CHANNEL_GROUPING",
    metrics={
        "GOAL_VALUE_ALL": ["SUM"],
        "NEW_USERS": ["AVG", "SUM"]
    },
    order_direction="DESC"
).sql)

ds7 = ga_ds.apply(
    sql=s,
    x_axis="CHANNEL_GROUPING",
    metrics={
        "GOAL_VALUE_ALL": ["SUM"],
        "NEW_USERS": ["AVG", "SUM"]
    },
    order_direction="DESC"
)
ds7.preview()
```

File: rasgotransforms/rasgotransforms/transforms/plot/plot.sql [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    {%- set bucket_count = num_buckets -%}
+{%- endif -%}
+
+-- if the axis is continuous or a date, do a line chart
+{% if axis_type in ['date', 'numeric'] -%}
+    WITH AXIS_RANGE AS (
+    -- Use a user-defined axis column to calculate the min & max of the axis (and buckets on the axis)
+    SELECT
+        {% if axis_type == 'date' -%}
+            MIN(DATE_PART(EPOCH_SECOND, {{ x_axis }}))-1 AS MIN_VAL
```
Comment (griffatrasgo) [inline rasgotransforms/rasgotransforms/transforms/plot/plot.sql:25]: Probably a dumb question a few weeks late: why cast to epoch here? Is this what the UI prefers and we punt timestamp formatting to it?

**Review by griffatrasgo — APPROVED** [review]
This is some pretty nice SQL, good job. It's hard AF to review jinja templates, so not promising there won't be rendering errors, but this looks directionally correct.

File: rasgotransforms/rasgotransforms/transforms/plot/plot.sql (hunk 1 above)
Comment (szaghikian-rasgo) [reply rasgotransforms/rasgotransforms/transforms/plot/plot.sql:25]: We're actually not sending them as epoch time in the query response - the UI is getting datetimes. The cast to epoch is necessary for some of the math that we do later in the query

File: rasgotransforms/rasgotransforms/transforms/plot/plot.sql [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        {%- endfor -%}
+    {%- endfor %}
+    FROM {{ source_table }}
+    {% if filter_statements is iterable -%}
+        {%- for filter_statement in filter_statements %}
+            {{ 'WHERE' if loop.first else 'AND' }} {{ filter_statement }}
+        {%- endfor -%}
+    {%- endif %}
+    GROUP BY {{ x_axis }}
+    {{ "ORDER BY " + x_axis + " " + order_direction if order_direction else '' }}
```
Comment (ChrisGriffithRASGO) [inline rasgotransforms/rasgotransforms/transforms/plot/plot.sql:98]: You probably want to add parns here to be clear of intent. Logically it works that either the ODER BY will be there or just an empty string as of now, but someone may read it incorrectly as  `"ORDER BY " + x_axis + " " + (order_direction if order_direction else '')`

```suggestion
    {{ ("ORDER BY " + x_axis + " " + order_direction) if order_direction else '' }}
```

**Pull request merged**